Changes in module 'nf-core/bwa/index'
--- modules/nf-core/bwa/index/main.nf
+++ modules/nf-core/bwa/index/main.nf
@@ -1,6 +1,9 @@
 process BWA_INDEX {
     tag "$fasta"
-    label 'process_single'
+    // NOTE https://bio-bwa.sourceforge.net/bwa.shtml#8
+    // requires 5.37N memory where N is the size of the database
+    memory { 5.5.B * fasta.size() }
+
 
     conda "${moduleDir}/environment.yml"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?

************************************************************
