Changes in module 'nf-core/pints/caller'
--- modules/nf-core/pints/caller/main.nf
+++ modules/nf-core/pints/caller/main.nf
@@ -1,6 +1,6 @@
 process PINTS_CALLER {
-    tag "$meta.id"
-    label 'process_medium'
+    tag "$meta.id" + "${chr_name ? ' | ' + chr_name : ''}"
+    label 'process_high'
 
     conda "${moduleDir}/environment.yml"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
@@ -8,7 +8,9 @@
         'biocontainers/pypints:1.1.8--pyh7cba7a3_0' }"
 
     input:
-    tuple val(meta), path(bams)
+    tuple val(meta), path(bams), path(bais)
+    val assay_type
+    each chr_name // optional
 
     output:
     tuple val(meta), path("*_divergent_peaks.bed")     , optional:true, emit: divergent_TREs
@@ -23,6 +25,7 @@
     script:
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
+    def chr = chr_name ? "--chromosome-start-with $chr_name" : ''
     // TODO handle bigwigs
     // def input_type  = ("${input[0]}".endsWith(".bam")) ? "--bam-file $input" :
     //                    ("$input".contains(".bw")) ? "--bw-pl ${input[0]} --bw-mn ${input[1]}" :
@@ -34,6 +37,8 @@
         --file-prefix $prefix \\
         --thread $task.cpus \\
         --dont-check-updates \\
+        --exp-type $assay_type \\
+        $chr \\
         $args
 
     cat <<-END_VERSIONS > versions.yml

************************************************************
