Changes in module 'nf-core/pints/caller'
--- modules/nf-core/pints/caller/main.nf
+++ modules/nf-core/pints/caller/main.nf
@@ -8,7 +8,8 @@
         'biocontainers/pypints:1.1.8--pyh7cba7a3_0' }"
 
     input:
-    tuple val(meta), path(bams)
+    tuple val(meta), path(bams), path(bais)
+    val assay_type
 
     output:
     tuple val(meta), path("*_divergent_peaks.bed")     , optional:true, emit: divergent_TREs
@@ -34,12 +35,13 @@
         --file-prefix $prefix \\
         --thread $task.cpus \\
         --dont-check-updates \\
+        --exp-type $assay_type \\
         $args
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
         python: \$(python --version | sed 's/Python //g')
-        pints: \$(pints_caller --version)
+        PINTS: \$(pints_caller --version)
     END_VERSIONS
     """
 }

************************************************************
